File: Unique-Binary-Search-Trees/NT.rs
pub fn num_trees(n: i32) -> i32 {
    use std::collections::HashMap;
    let mut m = HashMap::new();
    m.insert(0, 1);
    m.insert(1, 1);
    for i in 2..(n + 1) as usize {
        for j in 0..i {
            let temp = m.get(&i).unwrap_or(&0).clone();
            m.insert(
                i,
                *m.get(&j).unwrap_or(&0) * *m.get(&(i - j - 1)).unwrap_or(&0) + temp,
            );
        }
    }

    dbg!(&m);

    *m.get(&(n as usize)).unwrap() as i32
}

fn main() {
    dbg!(num_trees(3));
}

File: Unique-Binary-Search-Trees/NT.go
package main

import "fmt"

func num_trees(n int) int {
	table := map[int]int{
		0: 1,
		1: 1,
	}

	for i := 2; i <= n; i++ {
		for j := 0; j < i; j++ {
			table[i] += table[j] * table[i-j-1]
		}
	}
	fmt.Printf("%+v\n", table)
	return table[n]
}

func main() {
	println(num_trees(3))
}
